### PR TITLE
Safehandle finalization

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

File: glib/SafeObjectHandle.cs [hunk 1]
```diff
@@ -34,21 +37,28 @@ class SafeObjectHandle : SafeHandleZeroOrMinusOneIsInvalid
...
 		public static SafeObjectHandle Create (Object obj, IntPtr handle)
 		{
 			if (handle == IntPtr.Zero)
 				return Zero;
 
-			var safeHandle = new SafeObjectHandle (handle) {
-				tref = new ToggleRef (obj, handle)
-			};
+#if DEBUG_SAFEHANDLES
+			var safeHandle = new InternalCreateHandle (handle);
```
Comment (alanmcgovern) [inline glib/SafeObjectHandle.cs]: Shouldn't this be `var safeHandle = InternalCreateHandle?.Invoke (handle) ?? new SafeObjectHandle (handle);` ?

Comment (Therzok) [inline glib/SafeObjectHandle.cs]: I changed the way the logic is hooked up. I don't want to do that extra check.

File: glib/Object.cs [hunk 2]
```diff
@@ -32,69 +32,23 @@ namespace GLib {
...
-					if (tref != null) {
-						tref.Free ();
-						tref = null;
-					}
-				} catch (Exception e) {
-					Console.WriteLine ("Exception while disposing a " + this + " in Gtk#");
-					throw e;
-				}
+			try {
+				handle.Dispose ();
```
Comment (alanmcgovern) [inline glib/Object.cs]: We should be able to safely invoke `Dispose` twice - null check `handle` :)

File: glib/SafeObjectHandle.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+// Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+// Boston, MA 02111-1307, USA.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Win32.SafeHandles;
+
+namespace GLib
+{
+	public class SafeObjectHandle : SafeHandleZeroOrMinusOneIsInvalid
```
Comment (alanmcgovern) [inline glib/SafeObjectHandle.cs]: Should this be non-public?

File: bootstrap-2.12-debug-safehandles [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+# Run this to set configure.in up for an API version.
+
+DEBUG_SAFEHANDLES="-define:DEBUG_SAFEHANDLES"
```
Comment (alanmcgovern) [inline bootstrap-2.12-debug-safehandles]: Can we remove `DEBUG_SAFEHANDLES` as it's unused now?

File: glib/SafeObjectHandle.cs (hunk 3 above)
Comment (Therzok) [inline glib/SafeObjectHandle.cs]: Public for a reason. Subclassing so we can replace the safehandle and still have toggleref support

File: bootstrap-2.12-debug-safehandles (hunk 4 above)
Comment (Therzok) [inline bootstrap-2.12-debug-safehandles]: Yes

File: glib/Object.cs (hunk 2 above)
Comment (Therzok) [inline glib/Object.cs]: Sure!

**Comment by Therzok** [comment]
Should I use a simple object instead of a safehandle and implement finalizer/dispose? The current approach makes dispose queue on a timeout too. We'd also save a few bytes per object by just wrapping an IntPtr and not the additional safehandle states.

**Comment by alanmcgovern** [comment]
That'd make sense. We don't actually use SafeHandles correctly so it doesn't make a whole lot of sense. 

However, to go back to my original "should we make this private" question about th safehandle class - if we make these implementation details public we'll never be able to change them again. A normal finalizable object makes sense, but I think we should make it a non-public class ;)

**Comment by Therzok** [comment]
I have an idea of how to make it non-public and I will implement it.

**Comment by alanmcgovern** [comment]
Wanna squash and merge it?

**Pull request merged**